### DISCUSSION TITLE
(Feature request) Any chance to use different source of rgb-dem tiles?

### DISCUSSION THREAD

**Discussion by 2ktsch** [slutske22/leaflet-topography#4] [Ideas]
Hi! Is there any chance some parameters can be added (alongside mapbox api key) so that any source of rgb tiles (possibly even modifying offset and resolution depending on application) can be used?

**Comment by slutske22** [comment]
What other sources of rgb tiles are you thinking?  What do you mean exactly by offset and resolution?

**Reply by 2ktsch** [reply]
I was considering hosting my own, in the same tile server z/x/y.png format that seems to be standard.  By offset and resolution, I mean the -10000m that would be an rgb value of (0,0,0) and 0.1m (for each bitwise increment in the blue band) as Mapbox is doing (If I understood well-- they're just splitting a 24-bit integer into three octets and scaling by 10, right?).  I'm no javascript dev and haven't even checked out the source code... might do that later this week and submit a PR if I think it's something I can do, unless it's an easy mod you can do in a few minutes xD.

**Reply by slutske22** [reply]
I've been meaning to come back and do some code-cleanup, tweak a few things, so I added in this feature.  Check out the [custom-tile-url](https://github.com/slutske22/leaflet-topography/tree/custom-tile-url) branch.  Everything still seems to be working with the standard mapbox tiles there.  I don't really have any other tilelayers I know that I can use to test this out this new feature.  Do you know of any other rgb terrain layers besides mapbox?  I remember seeing a japanese tile server for these rgb dem tiles awhile ago, but I don't remember where.

Check out that branch, and check out the readme.  I added `tilesUrl` and `heightFunction`, which allows you to pick your own url for your tiles, and define your own height function.  Please clone the repo, and run `npm install` and `npm build`.  Then go into the example directory, run `npm install`, and `npm start`.  You can use that example to test things out.  Let me know what you find.  If its working, I can publish a new version.

**Reply by 2ktsch** [reply]
Thanks! Will do and get back here in a few days.  I'll also be generating a custom tile layer these days so we can test it with that --I'll make several variants of the same area to test the custom function.

**Reply by slutske22** [reply]
Ok I fixed a potential issue with using a custom tile url, but I have no way of testing it for now. LMK if you come up with some custome tilesets we can try.  Also, do you need this just for the central `getTopography` function?  Or do you also want to use custom tilesets / height functions for the Topolayer visualizations?

**Reply by 2ktsch** [reply]
Cool! I'm waiting for two versions of the same dataset I prepared to upload.  I'll post the link as soon as they're up.  The source is a 7m resolution DEM.  One of them has offset -50000 and 0.01m resolution and the other is the same as mapbox tileset (-10000 & 0.1m)  I'd be great if it worked with the topolayer visualizations!

![a chunk of PR](https://user-images.githubusercontent.com/13712904/122101501-6114fb00-cde2-11eb-9906-94c0c5386dca.png)
![zoomed in](https://user-images.githubusercontent.com/13712904/122101517-65d9af00-cde2-11eb-921e-311067bd72c6.png)
So trippy!

**Reply by slutske22** [reply]
Very cool.  yeah rgb-encoded dem is always trippy.  Totally incomprehensible though.  Hence `Topolayer`.  If we want to be able to customize custom rgb-dem tiles in the Topolayer visualizations, that may take some extra work.  I'll think about that one.  The `custom-tile-utl` branch should already work with the `getTopography` function.  LMK once those layers are up and I can mess with them.  Are they global, or local to a certain area?

**Reply by 2ktsch** [reply]
Here are the tiles! https://tileserver.tejidx.com/ folders are self-descriptive according to prev. post and I think they're not tms... check if Puerto Rico is the right side up when you do /z/x/y.png xD
They're only of Puerto Rico, Vieques and Culebra for now.  This 7m res one is a (manual) test for processing a 1m LIDAR-derived dataset (super zoom!) that was made in 2015-2016.
I'll be posting a repo for converting geotiff DEMs to tile layers once that is working.

**Reply by slutske22** [reply]
Ok I've got a quick [codesandbox](https://codesandbox.io/s/tedijx-tilelayer-testing-7m18b) going with the `rgb_50k100` layer.  What is the height function for this?  meaning `const height = (R, G, B) => ?`. Is it `-50000 + ((R * 256 * 256 + G * 256 + B) * 0.1)`?  What about for the `rgb_10k10` tiles?

**Reply by 2ktsch** [reply]
Almost- -50000 and 0.01  `height = (R, G, B) => -50000 + ( ( 256 * 256 * R + 256 * G + B) * 0.01)`
The 10k10 should work with the default function (-10000 and 0.1) `height = (R, G, B) => -10000 + ( ( 256 * 256 * R + 256 * G + B) * 0.1)`

**Reply by slutske22** [reply]
I'm pretty close to wrapping this up, but I'm stuck with getting a custom height function to work on a per-layer basis in the Topolayer visualizations.  You can check out my dillema [here](https://stackoverflow.com/questions/68029421/failed-to-execute-postmessage-on-worker-x-could-not-be-cloned), and read a bit more in the comments there.  If I can get past that issue, I think we can wrap this up and publish a new version which will enable the features you need.  If you have time and want to help me debug some web-worker stuff, please take a look.  Otherwise I'm not sure how long it will be before I have time / can figure this out.

**Reply by slutske22** [reply]
I am having CORS issues with your tilelayer when testing things out.  I set up my sample project to run on localhost:1234, and the tiles are fetched from there.  The tile fetching function triggers a CORS error:

<img width="1499" alt="Screen Shot 2021-06-22 at 11 47 10 AM" src="https://user-images.githubusercontent.com/52637023/122982328-9758fa00-d34f-11eb-80ff-47d9e3d3c91c.png">

As you can see, we have `referer: http://localhost:1234/`

When accessing the tiles directly in the browser, or using them in a tilelayer in codesandbox, it seems fine.  The referrer is actually `referer: https://tileserver.tejidx.com/rgb_50k100/10/320/`

Can you open up the referrer to 0.0.0.0:1234 (anything running on 1234), at least for testing?  I'm getting close and I'd love to test this out.  You can try to reproduce my issue by:

1. clone the repo
2. jump onto the `custom-tile-url` branch
3. `npm install` and `npm run build`
4. `cd example-ts`, and `npm i` and `npm start`
5. you should see the sample app running in the browser focused on puerto rico.  Look into the network tab, and you'll see a bunch of failed network requests due to CORS

I'm sort of stuck until your tileserver allows cross origin requests, at least for a narrow range of IP/ports required to dev this thing.

Thanks!

**Reply by 2ktsch** [reply]
whoops, sorry 'bout that! Should be fixed now-- open for anywhere ATM

**Comment by slutske22** [comment]
This is ready for testing.  You can try this out with the following:

1. `npm install leaflet-topography@next`
2. Use the `tilesUrl` option to define your custom tile url.  This can be used in liueu of the `token` option, either in the options passed to `getTopography` or a `TopoLayer`, or using the `configure` function.  In your case, I'd recommend using the `configure` function
3. Use the `heightFunction` option to define your custom heightFunction.  Same rules apply as for the `tilesUrl` option.

You can check out the [readme on the custom-tiles-url branch](https://github.com/slutske22/leaflet-topography/tree/custom-tile-url), which has a bit more writeup on `tilesUrl` and `heightFunction`.  There is still an issue using a custom `heightFunction` on a per-TopoLayer basis, but a workaround is described in the limitations section of the readme.

For a working example, follow these steps:

 1. Clone this repo, switch to the custom-tiles-url branch
 2. `cd example-ts`
 3. `npm install`
 4. `npm start`
 5. Go to localhost:1234
 6. You will see the demo app focused on puerto rico, using your custom tilelayer for the `getTopography` function and TopoLayers.

Please play around with this and let me know what you find.  If it seems like its all working, I'll publish this as a 0.2.0 version.

**Reply by 2ktsch** [reply]
Can't wait to try this out! Thanks! :D

**Reply by slutske22** [reply]
Ever get a chance to try this? If it’s working I’d like to publish to 0.2.0

**Reply by 2ktsch** [reply]
Sorry 'bout the delay-- yes! Haven't found any strange behavior.  Thanks again!
